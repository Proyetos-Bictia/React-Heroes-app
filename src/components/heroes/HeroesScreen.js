import React, { useMemo } from 'react'
import { useParams, Redirect } from 'react-router-dom'
import { getHeroesById } from '../../selectors/getHeroeById';

export const HeroesScreen = ({ history }) => {

    const { heroeId } = useParams();

    const hero = useMemo(() => getHeroesById(heroeId), [heroeId])

    if (!hero) {
        return <Redirect to="/" />
    }

    const {
        publisher,
        superhero,
        alter_ego,
        first_appearance,
        characters } = hero

    const hanldeReturn = () => {
        if (history.length <= 2) {
            history.push('/');
        } else {
            history.goBack();
        }
    }

    return (
        <div className="row mt-5">
            <div className="col-4 animate__animated animate__slideInLeft">
                <img
                    src={`../assets/heroes/${heroeId}.jpg`}
                    alt={superhero}
                    className="img-thumbnail"
                />
            </div>
            <div className="col-8">
                <h3>{superhero}</h3>
                <ul className="list-group list-group-flush">
                    <li className="list-group-item"><b>Alter ego: </b>{alter_ego}</li>
                    <li className="list-group-item"><b>Publisher: </b>{publisher}</li>
                    <li className="list-group-item"><b>First appearance: </b>{first_appearance}</li>
                </ul>
                <h5>Characters</h5>
                <p>{characters}</p>
                <button
                    className="btn btn-outline-primary"
                    onClick={hanldeReturn}
                >
                    Return
                </button>
            </div>
        </div>
    )
}
