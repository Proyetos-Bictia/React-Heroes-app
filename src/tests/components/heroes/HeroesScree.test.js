import '@testing-library/jest-dom';
import { mount, shallow } from 'enzyme';
import React from 'react';
import { MemoryRouter, Router, Route } from 'react-router-dom';
import { AuthContext } from '../../../auth/AuthContext';
import { types } from '../../../types/types';
import { HeroesScreen } from '../../../components/heroes/HeroesScreen';

describe('Pruebas en <HeroScreen />', () => {

    const historyMock = {
        length: 10,
        goBack: jest.fn(),
        push: jest.fn(),
    }


    test('debe de mostrar el redirect si no hay argumentos en el url ', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero']}>
                <HeroesScreen history={historyMock} />
            </MemoryRouter>
        );
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('Redirect').exists()).toBe(true)
    })

    test('debe de mostrarse correctamente si hay un hero', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-spider']}>
                <Route path="/hero/:heroeId" component={HeroesScreen} />
            </MemoryRouter>
        );
        expect(wrapper.find('.row').exists()).toBe(true)
        // console.log(wrapper.html());
    });

    test('debe de regresar a la pantalla anterior con PUSH', () => {
        const historyMock = {
            length: 1,
            goBack: jest.fn(),
            push: jest.fn(),
        }
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-spider']}>
                <Route
                    path="/hero/:heroeId"
                    component={(props) => <HeroesScreen history={historyMock} />}
                />
            </MemoryRouter>
        );
        wrapper.find('button').simulate('click');
        expect(historyMock.push).toHaveBeenCalledWith('/');
        expect(historyMock.goBack).not.toHaveBeenCalled()
    })

    test('debe de regresar a la pantalla anterior GOBACK', () => {
        const historyMock = {
            length: 3,
            goBack: jest.fn(),
            push: jest.fn(),
        }
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-spider']}>
                <Route
                    path="/hero/:heroeId"
                    component={(props) => <HeroesScreen history={historyMock} />}
                />
            </MemoryRouter>
        );
        wrapper.find('button').simulate('click');
        expect(historyMock.push).not.toHaveBeenCalled();
        expect(historyMock.goBack).toHaveBeenCalledTimes(1)
    })

    test('debe de llamar el redirect si el heroId no existe', () => {
        const historyMock = {
            length: 3,
            goBack: jest.fn(),
            push: jest.fn(),
        }
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-spider123131']}>
                <Route
                    path="/hero/:heroeId"
                    component={(props) => <HeroesScreen history={historyMock} />}
                />
            </MemoryRouter>
        );
        expect(wrapper).toEqual({})
    })

})
