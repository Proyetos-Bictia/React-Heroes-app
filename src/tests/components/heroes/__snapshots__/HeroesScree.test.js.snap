// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <HeroScreen /> debe de mostrar el redirect si no hay argumentos en el url  1`] = `""`;
